### PR TITLE
convert netty4.1 to java

### PR DESCRIPTION
**Pull request by shalk**
relate to #7195

### CODE REVIEW

**Comment by zeitlinger** [comment]
no groovy files are deleted

**Comment by shalk** [comment]
> no groovy files are deleted

Work in Progress

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+            channel.pipeline().addLast(new ClientHandler(result));
+            channel.writeAndFlush(request).get(10, TimeUnit.SECONDS);
+            result.get(10, TimeUnit.SECONDS);
+          });
+    } catch (Throwable e) {
+      thrownException = e;
+    }
+
+    List<AttributeAssertion> attributeAssertion =
+        Arrays.asList(
```
Comment (steverao) [inline instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java:134]: Suggest to use static import here.

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ConnectionSpanTest.java [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+  void testSuccessfulRequest() throws Exception {
+    URI uri = URI.create("http://localhost:" + server.httpPort() + "/success");
+    DefaultFullHttpRequest request = buildRequest("GET", uri, new HashMap<>());
+    int responseCode = testing.runWithSpan("parent", () -> sendRequest(request, uri));
+
+    assertThat(responseCode).isEqualTo(200);
+    testing.waitAndAssertTraces(
+        trace ->
+            trace
+                .hasSize(5)
```
Comment (steverao) [inline instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ConnectionSpanTest.java:118]: `hasSize` is unnecessary because `hasSpansSatisfyingExactly` will check the size. It seem there are many same problems, please modify them together.

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/ChannelFutureTest.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+          throws Exception {
+        counter.incrementAndGet();
+      }
+
+      @Override
+      public void operationComplete(Future future) throws Exception {
+        counter.incrementAndGet();
+      }
+    };
+  }
```
[the remark this answers is not in the document]
Comment (shalk) [reply instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/ChannelFutureTest.java:78]: But when i write 
```
  GenericFutureListener<Future<Void>> listener2 = newListener(counter);
    GenericFutureListener<ProgressiveFuture<Void>> listener3 = newProgressiveListener(counter);

    channel.closeFuture().addListeners(listener2, listener3);
```
```
channel.closeFuture().addListeners 
```
will have a warning `Unchecked generics array creation for varargs parameter `

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+  @BeforeEach
+  void setUp() {
+    server = new HttpClientTestServer(testing.getOpenTelemetry());
+    server.start();
+  }
+
+  @AfterEach
+  void tearDown() {
+    server.stop();
+  }
```
Comment (laurit) [inline instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java:61]: groovy version starts server once and uses it for all tests

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+            "parent1",
+            () -> {
+              Channel channel = b.connect("localhost", server.httpPort()).sync().channel();
+              channel.write(request);
+              channel.flush();
+              return channel;
+            });
+    // This is a cheap/easy way to block/ensure that the first request has finished and check
+    // reported spans midway through
+    // the complex sequence of events
```
Comment (laurit) [inline instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java:95]: merge these lines

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+    testing.waitAndAssertTraces(
+        trace ->
+            trace.hasSpansSatisfyingExactly(
+                span -> span.hasNoParent().hasName("parent1").hasKind(SpanKind.INTERNAL),
+                span -> span.hasParent(trace.getSpan(0)).hasKind(SpanKind.CLIENT),
+                span -> span.hasKind(SpanKind.SERVER).hasParent(trace.getSpan(1))));
+
+    // now run a second request through the same channel
```
Comment (laurit) [inline instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java:103]: perhaps it would be better to call `testing.clearData()` before the second request so that we wouldn't need to assert the first request again

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+        .isNotNull();
+  }
+
+  @Test
+  @DisplayName(
+      "when a traced handler is added from an initializer we still detect it and add our channel handlers")
+  void testAddInitializer() {
+    // This test method replicates a scenario similar to how reactor 0.8.x register the
+    // `HttpClientCodec` handler
+    // into the pipeline.
```
Comment (laurit) [inline instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java:195]: merge these lines

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+    protected void initChannel(Channel ch) throws Exception {
+      // This replicates how reactor 0.8.x add the HttpClientCodec
+      ch.pipeline().addLast("added_in_initializer", new HttpClientCodec());
+    }
+
+    @Override
+    @Deprecated
+    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+      super.exceptionCaught(ctx, cause);
+    }
```
Comment (laurit) [inline instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java:309]: why is this needed?

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+import java.util.concurrent.TimeUnit;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+class Netty41ClientAgentTest {
```
Comment (laurit) [inline instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java:45]: perhaps call it `Netty41ClientPipelineTest`

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+class Netty41ClientSslTest {
+
+  @RegisterExtension
+  static InstrumentationExtension testing = AgentInstrumentationExtension.create();
+
+  private HttpClientTestServer server;
+
+  private EventLoopGroup eventLoopGroup;
+
+  @SuppressWarnings("UnusedMethod")
```
Comment (laurit) [inline instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java:64]: why is this suppression needed?

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+  }
+
+  @AfterEach
+  void tearDown() throws ExecutionException, InterruptedException, TimeoutException {
+    server.stop().get(10, TimeUnit.SECONDS);
+    eventLoopGroup.shutdownGracefully();
+    System.clearProperty("otel.instrumentation.netty.connection-telemetry.enabled");
+    System.clearProperty("otel.instrumentation.netty.ssl-telemetry.enabled");
+  }
+
```
Comment (laurit) [inline instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java:106]: groovy version does these only once for class

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+        asList(
+            equalTo(ExceptionAttributes.EXCEPTION_MESSAGE, thrownException.getCause().getMessage()),
+            satisfies(ExceptionAttributes.EXCEPTION_STACKTRACE, v -> v.isNotNull()),
+            equalTo(
+                ExceptionAttributes.EXCEPTION_TYPE,
+                thrownException.getCause().getClass().getCanonicalName()));
+
+    testing.waitAndAssertTraces(
+        trace ->
+            trace.hasSpansSatisfyingExactly(
```
Comment (laurit) [inline instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java:143]: groovy test sorts RESOLVE/CONNECT/SSL handshake spans. Are you sure that you don't need to do this here? Won't this make the test flaky?

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+                                        ExceptionAttributes.EXCEPTION_TYPE,
+                                        v ->
+                                            v.isEqualTo(
+                                                SSLHandshakeException.class.getCanonicalName())),
+                                    satisfies(
+                                        ExceptionAttributes.EXCEPTION_STACKTRACE,
+                                        v -> v.isInstanceOf(String.class)),
+                                    satisfies(
+                                        ExceptionAttributes.EXCEPTION_MESSAGE, v -> v.isNotNull())))
+                        .hasAttributesSatisfyingExactly(
```
Comment (laurit) [inline instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java:186]: would `hasException(new SSLHandshakeException(null))` work?

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+                thrownException.getCause().getClass().getCanonicalName()));
+
+    testing.waitAndAssertTraces(
+        trace ->
+            trace.hasSpansSatisfyingExactly(
+                span ->
+                    span.hasName("parent")
+                        .hasNoParent()
+                        .hasStatus(StatusData.error())
+                        .hasEventsSatisfyingExactly(
```
Comment (laurit) [inline instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java:148]: you probably could replace this with `hasException(thrownException)`, you'll have to unwrap the exception the same way groovy test does it

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+        new DefaultFullHttpRequest(
+            HttpVersion.HTTP_1_1, HttpMethod.GET, uri.getPath(), Unpooled.EMPTY_BUFFER);
+    request.headers().set(HttpHeaderNames.HOST, uri.getHost() + ":" + uri.getPort());
+
+    Throwable thrownException = null;
+    try {
+      testing.runWithSpan(
+          "parent",
+          () -> {
+            Channel channel = bootstrap.connect(uri.getHost(), uri.getPort()).sync().channel();
```
Comment (laurit) [inline instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java:123]: groovy test closes the channel as part of the cleanup at the end of the test. Search for `AutoCleanupExtension` in instrumentation tests, you can use it to close the channel at the end of the test.

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/ChannelFutureTest.java [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+class ChannelFutureTest {
+
+  @RegisterExtension
+  static final InstrumentationExtension testing = AgentInstrumentationExtension.create();
+
+  // regression test for
+  // https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/2705
+  @DisplayName("should clean up wrapped listeners")
+  @Test
+  @SuppressWarnings({"unchecked"})
```
Comment (laurit) [inline instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/ChannelFutureTest.java:32]: for a single suppression it is more common to use `@SuppressWarnings("unchecked")`

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ConnectionSpanTest.java [hunk 17]
```diff
@@ -0,0 +1,232 @@
...
+    URI uri = URI.create("http://localhost:" + PortUtils.UNUSABLE_PORT);
+    DefaultFullHttpRequest request = buildRequest("GET", uri, new HashMap<>());
+    Throwable thrownException = null;
+    try {
+      testing.runWithSpan("parent", () -> sendRequest(request, uri));
+    } catch (Throwable e) {
+      thrownException = e;
+    }
+
+    Throwable finalThrownException = thrownException;
```
Comment (jaydeluca) [inline instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ConnectionSpanTest.java:156]: You could replace this block with:

```suggestion
    Throwable finalThrownException =
        catchThrowable(() -> testing.runWithSpan("parent", () -> sendRequest(request, uri)));
```

And then below in the assertions you can get rid of the `.hasEventsSatisfyingExactly` blocks and replace them with

```java
.hasException(finalThrownException),
```

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java (hunk 4 above)
Comment (shalk) [reply instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java]: done

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java (hunk 5 above)
Comment (shalk) [reply instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java]: done

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java (hunk 6 above)
Comment (shalk) [reply instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java]: done

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java (hunk 7 above)
Comment (shalk) [reply instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java]: done

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java (hunk 8 above)
Comment (shalk) [reply instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java]: removed, i think it  is not need

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java (hunk 9 above)
Comment (shalk) [reply instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientAgentTest.java]: done

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java (hunk 10 above)
Comment (shalk) [reply instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java]: removed

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java (hunk 11 above)
Comment (shalk) [reply instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java]: done

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java (hunk 12 above)
Comment (shalk) [reply instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java:136]: the test works fine

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java (hunk 13 above)
Comment (shalk) [reply instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java]: it works.

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java (hunk 14 above)
Comment (shalk) [reply instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java]: done

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java (hunk 15 above)
Comment (shalk) [reply instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java:122]: done

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java (hunk 12 above)
Comment (jaydeluca) [reply instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java:134]: I think the concern is that it might work fine most of the time, but it could also become flaky and result in sporadic failures

Comment (shalk) [reply instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java:134]: > I think the concern is that it might work fine most of the time, but it could also become flaky and result in sporadic failures

let's fix it more stable.

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java [hunk 18]
```diff
@@ -0,0 +1,238 @@
...
+    return bootstrap;
+  }
+
+  @BeforeAll
+  static void setUp() {
+    server = new HttpClientTestServer(testing.getOpenTelemetry());
+    server.start();
+    eventLoopGroup = new NioEventLoopGroup();
+    System.setProperty("otel.instrumentation.netty.connection-telemetry.enabled", "true");
+    System.setProperty("otel.instrumentation.netty.ssl-telemetry.enabled", "true");
```
Comment (laurit) [inline instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java:96]: Why do you need to set these here? These are already set in gradle build file, groovy tests don't seem to have these.

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java [hunk 19]
```diff
@@ -0,0 +1,238 @@
...
+  }
+
+  @AfterAll
+  static void tearDown() throws ExecutionException, InterruptedException, TimeoutException {
+    server.stop().get(10, TimeUnit.SECONDS);
+    eventLoopGroup.shutdownGracefully();
+    System.clearProperty("otel.instrumentation.netty.connection-telemetry.enabled");
+    System.clearProperty("otel.instrumentation.netty.ssl-telemetry.enabled");
```
Comment (laurit) [inline instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java:104]: same as the previous

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java (hunk 12 above)
Comment (laurit) [reply instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java]: Sorry, after thinking about this a bit more could you change it back to `hasSpansSatisfyingExactly`. I think the underlying issue that caused the spans to appear in different order was probably resolved so sorting the spans shouldn't be necessary. If it turns out to be necessary we can fix it later.

**Review by jaydeluca — APPROVED** [review]

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java (hunk 12 above)
Comment (shalk) [reply instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/Netty41ClientSslTest.java]: I will revert it back

File: instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/ChannelPipelineTest.java [hunk 20]
```diff
@@ -0,0 +1,212 @@
...
+    channelPipeline.addAfter("http", "noop", noopHandler);
+
+    // instrumentation handler is between with http and noop;
+    assertThat(channelPipeline.first()).isEqualTo(handler);
+    assertThat(channelPipeline.last()).isEqualTo(noopHandler);
+
+    channelPipeline.removeLast();
+
+    // http and instrumentation handlers will be remained;
+    assertThat(channelPipeline.first()).isEqualTo(handler);
```
[the remark this answers is not in the document]
Comment (shalk) [reply instrumentation/netty/netty-4.1/javaagent/src/test/java/io/opentelemetry/javaagent/instrumentation/netty/v4_1/ChannelPipelineTest.java:187]: done, add more assertions

**Review by laurit — APPROVED** [review]
